extensions/sql/policy/store/src/main/java/org/eclipse/dataspaceconnector/sql/policy/store/SqlPolicyStore.java

```diff
@@ -0,0 +1,190 @@
+import java.util.Objects;
+import java.util.stream.Stream;
+import javax.sql.DataSource;
+
+import static org.eclipse.dataspaceconnector.sql.SqlQueryExecutor.executeQuery;
+
+public class SqlPolicyStore implements PolicyStore {
+
+    private final DataSourceRegistry dataSourceRegistry;
+    private final String dataSourceName;
+    private final TransactionContext transactionContext;
+    private final TypeManager typeManager;
+    private final SqlPolicyStoreStatements sqlPolicyStoreStatements;
+
+    public SqlPolicyStore(DataSourceRegistry dataSourceRegistry, String dataSourceName, TransactionContext transactionContext,
+                          TypeManager typeManager, SqlPolicyStoreStatements sqlPolicyStoreStatements) {
+        this.dataSourceRegistry = Objects.requireNonNull(dataSourceRegistry);
+        this.dataSourceName = Objects.requireNonNull(dataSourceName);
+        this.transactionContext = Objects.requireNonNull(transactionContext);
+        this.typeManager = Objects.requireNonNull(typeManager);
+        this.sqlPolicyStoreStatements = Objects.requireNonNull(sqlPolicyStoreStatements);
+    }
+
+    @Override
+    public @Nullable Policy findById(String id) {
+        try {
+            return single(executeQuery(getConnection(), this::mapResultSet, sqlPolicyStoreStatements.getSqlFindByClauseTemplate(), id));
+        } catch (SQLException exception) {
+            throw new EdcPersistenceException(exception);
+        }
+    }
+
+    @Override
+    public Stream<Policy> findAll(QuerySpec querySpec) {
+        Objects.requireNonNull(querySpec);
+        var query = sqlPolicyStoreStatements.getSqlFindClauseTemplate();
+
+        try (var connection = getConnection()) {
+            return executeQuery(connection, this::mapResultSet, query, querySpec.getLimit(), querySpec.getOffset()).stream();
+        } catch (Exception exception) {
+            throw new EdcPersistenceException(exception);
+        }
+    }
+
+    @Override
+    public void save(Policy policy) {
+        Objects.requireNonNull(policy);
+        transactionContext.execute(() -> {
+            if (findById(policy.getUid()) != null) {
+                update(policy);
+            } else {
+                insert(policy);
+            }
+        });
+    }
+
+    @Override
+    public Policy deleteById(String policyId) {
+        Objects.requireNonNull(policyId);
+        return transactionContext.execute(() -> {
+            try (var connection = getConnection()) {
+                var entity = findById(policyId);
+                if (entity != null) {
+                    executeQuery(connection, sqlPolicyStoreStatements.getSqlDeleteClauseTemplate(), policyId);
+                }
+                return entity;
+            } catch (Exception e) {
+                throw new EdcPersistenceException(e.getMessage(), e);
+            }
+        });
+    }
+
+    private void insert(Policy policy) {
+        try (var connection = getConnection()) {
```

Comment:
`insert`/`update`: they also should execute their code in a transaction. I know they are private, and that the `save` method already creates one, but that would just "attach" `save`/`update` to the existing transaction and would increase robustness, should we ever call those two from another method.